### PR TITLE
Cluster-autoscaler scale down impl

### PR DESCRIPTION
**Pull request by mwielgus**
cc: @piosz @fgrzadkowski

### CODE REVIEW

**Comment by piosz** [comment]
ping me when it's ready for the review

**Comment by mwielgus** [comment]
@piosz  Ping

File: cluster-autoscaler/simulator/cluster.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	"k8s.io/kubernetes/pkg/api/resource"
+	kube_client "k8s.io/kubernetes/pkg/client/unversioned"
+	cmd "k8s.io/kubernetes/pkg/kubectl/cmd"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// Nodes with utilization below this are considered unused and mayb be subject to scale down.
```
Comment (piosz) [inline cluster-autoscaler/simulator/cluster.go]: typo: mayb

File: cluster-autoscaler/simulator/cluster.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	unusedThreshold = float64(0.5)
+)
+
+// FindNodeToRemove finds a node that can be removed.
+func FindNodeToRemove(nodes []*kube_api.Node, pods []*kube_api.Pod, client *kube_client.Client) (*kube_api.Node, error) {
+
```
Comment (piosz) [inline cluster-autoscaler/simulator/cluster.go]: remove this line

File: cluster-autoscaler/simulator/cluster.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	nodeNameToNodeInfo := schedulercache.CreateNodeNameToInfoMap(pods)
+
+	for _, node := range nodes {
+		nodeInfo, found := nodeNameToNodeInfo[node.Name]
+		if !found {
+			glog.Errorf("Node info for %s not found", node.Name)
+			continue
+		}
+		reservation, err := calculateReservation(node, nodeInfo, kube_api.ResourceCPU)
```
Comment (piosz) [inline cluster-autoscaler/simulator/cluster.go]: TODO here

File: cluster-autoscaler/simulator/cluster.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	// Nodes with utilization below this are considered unused and mayb be subject to scale down.
+	unusedThreshold = float64(0.5)
+)
+
+// FindNodeToRemove finds a node that can be removed.
+func FindNodeToRemove(nodes []*kube_api.Node, pods []*kube_api.Pod, client *kube_client.Client) (*kube_api.Node, error) {
+
+	nodeNameToNodeInfo := schedulercache.CreateNodeNameToInfoMap(pods)
+
+	for _, node := range nodes {
```
Comment (piosz) [inline cluster-autoscaler/simulator/cluster.go]: add TODO: iterate from least utilized nodes

File: cluster-autoscaler/simulator/cluster.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		if findProblems == nil {
+			return node, nil
+		} else {
+			glog.Infof("Node %s is not suitable for removal %v", node.Name, err)
+		}
+	}
+	return nil, nil
+}
+
+func calculateReservation(node *kube_api.Node, nodeInfo *schedulercache.NodeInfo, resourceName kube_api.ResourceName) (float64, error) {
```
Comment (piosz) [inline cluster-autoscaler/simulator/cluster.go]: I'm pretty sure this logic is implemented somewhere in the kubelet, so it'd great to reuse it (not in this PR).

**Comment by piosz** [comment]
LGTM, just a few nits

File: cluster-autoscaler/simulator/cluster.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/simulator/cluster.go]: Done

File: cluster-autoscaler/simulator/cluster.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/simulator/cluster.go]: Done.

File: cluster-autoscaler/simulator/cluster.go (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/simulator/cluster.go]: Done.

File: cluster-autoscaler/simulator/cluster.go (hunk 5 above)
Comment (mwielgus) [inline cluster-autoscaler/simulator/cluster.go]: Ack.

**Pull request merged**